### PR TITLE
Move pass manager passes into header, convert ExpandExternArrayCalls

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Move pass manager passes into header, convert ExpandExternArrayCalls (#20014)

Convert the `expandExternArrayCalls` pass into a pass that is run
by the new pass manager.

Create a new header `pass-manager-passes.h` to store all the passes.
They are listed in approximate execution order, but that may change
as things are run out of order, and some passes are already run
multiple times.

In the future we might like to dedicate a header per pass (and clean
up source files accordingly), but I think this is a fine substitute
in the interim.

TESTING

- [x] `ALL` on `linux64`, `standard`

Reviewed by @lydia-duncan. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Hi, @lydia-duncan, would you be willing to review the `ExpandExternArrayCalls` pass? I think it turned out relatively straight forward as far as breaking it up into predicate + work.

**Review by lydia-duncan — APPROVED** [review]
Looks good!  Just called out a typo and a couple of places that didn't look like how I was expecting but are probably fine

File: compiler/include/pass-manager-passes.h [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+  isFormalArray(ArgSymbol* formal, UnresolvedSymExpr** outType=nullptr);
+
+  static bool isRetExprVoid(BlockStmt* retExprType);
+};
+
+/**
+  Flatten all classes so they are defined at module scope.
+*/
+class FlattenClasses : public PassT<TypeSymbol*> {
+  void process(TypeSymbol* ts) override;
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply compiler/include/pass-manager-passes.h:51]: In what ways are it different? Do you mean that it doesn't have a `shouldProcess`?

File: compiler/include/pass-manager-passes.h [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  bool shouldProcess(FnSymbol* fn) override;
+  void process(FnSymbol* fn) override;
+};
+
+/**
+  Pull out global constants from loop runs in order to reduce loads.
+*/
+class LocalizeGlobals : public PassT<FnSymbol*> {
+ public:
+  void process(FnSymbol* fn) override;
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply compiler/include/pass-manager-passes.h:131]: Good question. The default for `shouldProcess` is to always return `true`. Let me go read the pass.

File: compiler/include/pass-manager-passes.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (lydia-duncan) [reply compiler/include/pass-manager-passes.h:51]: Yeah, that's what I mean.  And also slightly that it doesn't declare it as "public", but like, I want to say that's the default anyways?

[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply compiler/include/pass-manager-passes.h:51]: D'oh! Nah the default is private, great catch ty

File: compiler/include/pass-manager-passes.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply compiler/include/pass-manager-passes.h:131]: Yep, always returns true.

File: compiler/include/pass-manager-passes.h [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+*/
+class ReturnStarTuplesByRefArgs1 : public PassT<FnSymbol*> {
+  bool shouldProcess(FnSymbol* fn) override;
+  void process(FnSymbol* fn) override;
+};
+
+/**
+  The second pass may be run immediately after the first pass, and
+  adjusts all calls which get star tuple fields (to handle the
+  fact that they are now all acessed via ref).
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply compiler/include/pass-manager-passes.h:158]: Thanks!

**Pull request merged**